PUL-17 RBAC and tests.

sriniadhiwork opened the pull request:
1.  JWTUserconverter code needs to be enhanced and tested for Orgs extraction
2.  Pulse-user/saml references need to be cleaned, have circumvented some places by adding null check, but need to discuss the proper replacement.
3.  Few more tests pending.
5. Have circumvented a lazy loading issue by making it eager, but need to discuss the right approach. See AlternateCareFacilityEntity.java,  column alternate_care_facility_id, onetomany.
6. Other integration issues, related to JWT parsing [portlet uses io.jwts which allows maps to be stored, but RsaJose4 has only List<Strings> (looks like).

sriniadhiwork committed:
PUL-17

1. Service updated to perform ROLE checks
2. Broker layer uses spring security expressions to configure.
3. Controller tests to test the security from service - need to enhance.
4. ACF is retrieved using the liferay id in JWT from the database,
instead of from the JWT itself.

sriniadhiwork committed:
tests refactored; also fixed issues due to dto -> domain conversions
setting the liferay ids. questions: 1. FetchType.LAZY in
AlternateCareFacilityEntity causing issues - changing to
FetchType.FetchType.EAGER, needs investigation. 2. Need to make sure
that the acf fetched from the jwt vs  acf fetched from the db pointed to
by the liferay_acf_id in the jwt.  3. some tests need review/more setup
to pass. 4. merge with the latest code. 5. Review the jwt and other
fields.

sriniadhiwork committed:
Merging the latest changes from pulse-us to PUL-17

resolved the conflicts

sriniadhiwork committed:
commenting out the "Orgs" extraction.

sriniadhiwork committed:
minor fix: formatting, and blocking search for org_admin

sriniadhiwork commented:
The conflict can be resolved by keeping the original, as my changes are the same (just the order and formatting has changed.)

andlar commented on pulse/broker/src/main/java/gov/ca/emsa/pulse/service/AlternateCareFacilityService.java:
```diff
@@ -60,6 +67,7 @@ public AlternateCareFacility getById(@PathVariable("acfId") Long acfId) {
 
 	@ApiOperation(value = "Create a new ACF")
 	@RequestMapping(value = "/create", method = RequestMethod.POST)
+	@Secured({"ROLE_ADMIN", "ROLE_ORG_ADMIN"})
```
ROLE_PROVIDER needs to be able to create ACFs as well. If they're the first one to log in with a given ACF they'll automatically create it

andlar commented on pulse/broker/src/main/java/gov/ca/emsa/pulse/service/AlternateCareFacilityService.java:
```diff
@@ -78,19 +92,21 @@ public AlternateCareFacility create(@RequestBody(required=true) AlternateCareFac
 
 	@ApiOperation(value = "Edit an existing ACF")
 	@RequestMapping(value = "/{acfId}/edit", method = RequestMethod.POST)
+	@PreAuthorize("hasPermissionForAcf(#acfId)")
```
Will ROLE_PROVIDER return true for this check? Though they can create ACFs they can't edit them; this might also need a "has role: ROLE_ADMIN or ROLE_ORG_ADMIN"

andlar commented on pulse/broker/src/main/java/gov/ca/emsa/pulse/service/AlternateCareFacilityService.java:
```diff
@@ -78,19 +92,21 @@ public AlternateCareFacility create(@RequestBody(required=true) AlternateCareFac
 		
 		CommonUser user = UserUtil.getCurrentUser();
 		//auditManager.addAuditEntry(QueryType.EDIT_ACF, user.getAcf() + "/edit", user.getSubjectName());
-		if(user.getAcf() == null) {
```
Can just delete this commented out section. No longer needed, and git is great about history

andlar commented on pulse/broker/src/main/java/gov/ca/emsa/pulse/service/PulseUserService.java:
```diff
@@ -30,6 +31,7 @@
 	
 	@ApiOperation(value = "Create a new User")
 	@RequestMapping(value = "/create", method = RequestMethod.POST)
+	@Secured({"ROLE_ADMIN", "ROLE_ORG_ADMIN"})
```
I'm pretty sure ROLE_PROVIDER should be able to be a PULSE user too

andlar commented on pulse/service/src/main/java/gov/ca/emsa/pulse/service/PulseUserService.java:
```diff
@@ -38,6 +40,7 @@
 
 	@ApiOperation(value = "Create a new User")
 	@RequestMapping(value = "/create", method = RequestMethod.POST)
+	@Secured({"ROLE_ADMIN", "ROLE_ORG_ADMIN"})
```
ROLE_PROVIDER should be able to be a PULSE User too

andlar commented on pulse/auth/src/main/java/gov/ca/emsa/pulse/auth/authentication/JWTUserConverterImpl.java:
```diff
@@ -68,15 +75,32 @@ public User getAuthenticatedUser(String jwt) throws JWTValidationException {
+			/*String orgsString =  (String) validatedClaims.get("Orgs");
+			HashMap<String, Long> orgsInfo = null;
+			
+			try {
```
What does have this left out mean still needs to be done?

andlar reviewed (changes requested):
I just created a MR with some documentation for how to configure Eclipse to use the coding standards & conventions I'd like to have the team work from: https://git.ainq.com/pulse/documentation/blob/eclipse/us/docs/eclipse/eclipse_setup.md

You don't need to update this PR to meet those standards immediately, but if you want to that'd be great. Theoretically, it should remove a lot off the diffs that are in this PR where spaces became tabs and such.

Regarding the "cannot merge now" warning, you should be able to fix that by doing the merge on your local system then pushing a commit to this branch that has the updated merge. The PR can't be merged until that's done, so it wouldn't hurt to do it now

sriniadhiwork committed:
implemented reviews from pulse-us/api/pull/3

The checkstyle/formatting changes are going to set the diffs but
hopefully can be ignored by adding ?w=1 in the github url. Other updates
- the pending JWTUser for Orgs has been implemented.

sriniadhiwork commented:
Working on resolving the conflicts. Also use https://github.com/pulse-us/api/pull/3/files?w=1 to view without the whitespace changes.

sriniadhiwork committed:
Merge branch 'development' of https://github.com/pulse-us/api into PUL-17

# Conflicts:
#	pulse/auth/src/test/java/gov/ca/emsa/pulse/auth/authentication/JWTUserConverterImplTest.java
#	pulse/broker/src/main/java/gov/ca/emsa/pulse/broker/dto/DomainToDtoConverter.java
#	pulse/broker/src/main/java/gov/ca/emsa/pulse/broker/manager/impl/DocumentQueryService.java
#	pulse/broker/src/test/resources/application-test.properties.template
#	pulse/service/src/main/java/gov/ca/emsa/pulse/service/DocumentQueryService.java
